src/harness/reference_models/iap/iap.py

```diff
@@ -0,0 +1,623 @@
+    It is updated in the format of {latitude : {longitude : [ asas_p(mW/IAPBW),.. ]}}
+    and aggregate_interference instance of class type AggregateInterferenceOutputFormat.
+    It is updated in the format of {latitude : { longitude : [a_p(mW/IAPBW),..] }}
+  """
+  
+  # Get all the grants inside neighborhood of the protection entity
+  grants_inside = interf.findGrantsInsideNeighborhood(
+                    grant_objects, protection_point, protection_ent_type)
+   
+  if len(grants_inside) > 0:
+    # Get protection constraint of protected entity
+    protection_constraint = interf.ProtectionConstraint(latitude=protection_point[1],
+                               longitude=protection_point[0], low_frequency=low_freq,
+                               high_frequency=high_freq, entity_type=protection_ent_type)
+
+    # Get all the grants fully/partially overlap with protection point,p
+    neighborhood_grants = interf.findOverlappingGrants(
+                            grants_inside, protection_constraint)
+
+    num_unsatisfied_grants = len(neighborhood_grants)
+    
+    # Get number of grants, IAP protection threshold and fairshare per channel 
+    if num_unsatisfied_grants > 0:
+      # algorithm to calculate interference within IAPBW using EIRP obtained for all 
+      # the grants through application of IAP
+
+      # Initialize list of aggregate interference from all the grants 
+      # ( including grants from managing and peer SAS )
+      aggr_interf = [0] * len(channels)
+      # Initialize list of aggregate interference from all the grants 
+      # ( including only grants from managing SAS )
+      asas_interf = [0] * len(channels)
+
+      num_unsatisfied_grants_channel = []
+      iap_threshold_channel = []
+      fairshare_channel = []
+
+      for index, channel in enumerate(channels):
+        # Get protection constraint over 5MHz channel range
+        channel_constraint = interf.ProtectionConstraint(latitude=protection_point[1],
+                               longitude=protection_point[0], low_frequency=channel[0],
+                               high_frequency=channel[1], entity_type=protection_ent_type)
+
+        # Identify CBSD grants overlapping with frequency range of the protection entity 
+        # in the neighborhood of the protection point and channel
+        channel_neighborhood_grants = interf.findOverlappingGrants(
+                                        neighborhood_grants, channel_constraint)
+
+        num_unsatisfied_grants_channel.append(len(channel_neighborhood_grants))
+        
+        # Set interference quota for protection_point, channel combination
+        # Calculate Roll of Attenuation for ESC Sensor Protection entity
+        if protection_ent_type is interf.ProtectedEntityType.ESC:
+          if channel[0] >= 3650.e6:
+            center_freq = (channel[0] + channel[1]) / 2
+            iap_threshold_channel.append(interf.dbToLinear(interf.linearToDb(threshold) + 
```

Comment:
This logic has still not been addressed, despite pinpointing this several time. 
I guess it is my fault since I said I would look at this in more details. So here it is:

we dont think that "increasing threshold" and "decreasing interference" are equivalent: when a grant is satisfied, it is because "interference < remaining_threshold / num_unsatisfied_grants". So to get the same result between the 2 approaches, ones would have to increase the threshold by num_unsatisfied_grants * (amount that should be added to interferemce). This is not doable at the beginning of the process as the value of Num_unsatisfie_grants changes as time goes in the loop.

So I think we should change instead the intreference calculation for ESC slightly out of band, to apply the roll off to them instead, and keep the threshold as it is. 
Ideally the application of that roll off should be done as part of the computeInterference routine (but probably this means adding the channel info to that one)

Tell me if you think otherwise.

Reply:
We have reviewed the spec against our interception. It seems that we need to have a discussion on this. Would you be available for an offline conference discussion ?

Reply:
Serge, here is our understanding

Recalling
R2-SGN-25 SAS ESC Sensor Protection 
a. SASs shall manage CBRS interference for all ESC sensors that require protection such that the aggregate mean interference at the reference ESC filter output of the protected sensor in 3550-3700 MHz does not exceed - 109 dBm/MHz. The reference ESC filter has 0.5 dB insertion loss in the passband 
b. For initial certification, the SAS shall treat Category B CBSDs operating within the frequency range 3650-3680 MHz, and Category A CBSDs operating within the frequency range 3650-3660 MHz, as co-channel to ESCs and apply the same protection described above, after assuming a straight line 1 dB per MHz ESC reference filter roll-off from 3650-3680 MHz 

Our interpretation of the requirement is that the IAP calculations should be applied without modification after adjusting for the ESC reference filter roll-off.  This is accomplished by adjusting the threshold (Q) to compensate for the roll-off, i.e., increasing it by an amount corresponding to the roll-off.  Then the IAP algorithm (same protection described above) can be applied.  Modifying the individual grant interference contributions based on the rolloff would be a modification of the "protection described above."  Q represents the acceptable amount of aggregate interference at the ESC filter input.  The filter then attenuates the aggregate by the rolloff.

Reply:
Looks like these two approaches are equivalent mathematically. So it's fine to leave as it is.

Reply:
I think your interpretation works as well. But the problem is that in this reference implementation, after IAP is completed, it does not attenuates the aggregate by the ESC rolloff.  -- Robin (from Google).

Reply:
Agreed, Updated the code for ESC   rolloff calculation